parachain-system/src/lib.rs

```diff
@@ -443,6 +435,119 @@ decl_module! {
 }
 
 impl<T: Config> Module<T> {
+	/// Process all inbound downward messages relayed by the collator.
+	///
+	/// Checks if the sequence of the messages is valid, dispatches them and communicates the number
+	/// of processed messages to cumulus via a storage update.
+	fn process_inbound_downward_messages(
+		vfp: &PersistedValidationData,
+		downward_messages: Vec<InboundDownwardMessage>,
+	) -> DispatchResult {
+		let mut running_mqc_head = LastDmqMqcHead::get();
+		let dm_count = downward_messages.len() as u32;
+		for downward_message in downward_messages {
+			running_mqc_head = BlakeTwo256::hash_of(&(
+				running_mqc_head,
+				downward_message.sent_at,
+				BlakeTwo256::hash_of(&downward_message.msg),
+			));
+			T::DownwardMessageHandlers::handle_downward_message(downward_message);
+		}
+
+		// After hashing each message in the message queue chain submitted by the collator, we should
+		// arrive to the MQC head provided by the relay chain.
+		ensure!(running_mqc_head == vfp.dmq_mqc_head, Error::<T>::DmpMqcMismatch);
+
+		LastDmqMqcHead::put(running_mqc_head);
+		// Store the processed_downward_messages here so that it will be accessible from
+		// PVF's `validate_block` wrapper and collation pipeline.
+		storage::unhashed::put(well_known_keys::PROCESSED_DOWNWARD_MESSAGES, &dm_count);
+
+		Ok(())
+	}
+
+	/// Process all inbound horizontal messages relayed by the collator.
+	///
+	/// This is similar to `process_inbound_downward_messages`, but works on multiple inbound
+	/// channels.
+	fn process_inbound_horizontal_messages(
+		vfp: &PersistedValidationData,
+		horizontal_messages: BTreeMap<ParaId, Vec<InboundHrmpMessage>>,
+	) -> DispatchResult {
+		// First prepare horizontal messages for a more convenient processing:
+		//
+		// instead of a mapping from a para to a list of inbound HRMP messages, we will have a list
+		// of tuples `(sender, message)` first ordered by `sent_at` (the relay chain block number
+		// in which the message hit the relay-chain) and second ordered by para id ascending.
+		//
+		// The messages will be dispatched in this order.
+		let mut horizontal_messages = horizontal_messages
+			.into_iter()
+			.flat_map(|(sender, channel_contents)| {
+				channel_contents
+					.into_iter()
+					.map(move |message| (sender, message))
+			})
+			.collect::<Vec<_>>();
+		horizontal_messages.sort_by(|a, b| {
+			// first sort by sent-at and then by the para id
+			match a.1.sent_at.cmp(&b.1.sent_at) {
+				cmp::Ordering::Equal => a.0.cmp(&b.0),
+				ord => ord,
+			}
+		});
+
+		let last_mqc_heads = LastHrmpMqcHeads::get();
+		let mut running_mqc_heads = BTreeMap::new();
```

Comment:
When a parachain didn't send us a message in this round, don't we currently delete the last head? Don't we require this when we receive the next message in the future?

Reply:
Fixed by https://github.com/paritytech/cumulus/pull/308/commits/bf6fff56206c3a0f40c9f0549ca3e2a97b80f50b, good catch